compiler/src/dmd/errors.d

```diff
@@ -35,6 +35,133 @@ enum ErrorKind
     message,
 }
 
+// Struct for SARIF Tool Information
+struct ToolInformation {
+    string name;
+    string toolVersion;
+
+    string toJson() nothrow {
+        return `{"name": "` ~ name ~ `", "version": "` ~ toolVersion ~ `"}`;
+    }
+}
+
+// Struct for SARIF Result
+struct Result {
+    string ruleId;  // Rule identifier
+    string message;  // Error message
+    PhysicalLocation location;  // Location where the error occurred
+
+    string toJson() nothrow {
+        return `{"ruleId": "` ~ ruleId ~ `", "message": "` ~ message ~ `", "location": ` ~ location.toJson() ~ `}`;
+    }
+}
+
+// Struct for Physical Location of the Error
+struct PhysicalLocation {
```

Comment:
This looks very similar to https://github.com/dlang/dmd/blob/master/compiler/src/dmd/location.d#L235

It would probably be best if you just did a function that takes a `SourceLoc ` and returns a JSON string representation for it.

Reply:
Thanks! I’ll switch to using `OutBuffer` with `ExtractSlice` and create a stateful function for `SourceLoc` to return a JSON string, similar to `location.d`.

Reply:
(I think you resolved the wrong conversation here)

Reply:
Sorry for that. Please review- https://github.com/dlang/dmd/pull/16986/commits/d8d93295cc5ad5bcbff0a33c9ee9d6940bf7bb38